modules/ghlint-rules/src/main/kotlin/net/twisterrob/ghlint/rules/RequiredPermissionsRule.kt

```diff
@@ -0,0 +1,229 @@
+package net.twisterrob.ghlint.rules
+
+import net.twisterrob.ghlint.model.Access
+import net.twisterrob.ghlint.model.Permission
+import net.twisterrob.ghlint.model.Scope
+import net.twisterrob.ghlint.model.WorkflowStep
+import net.twisterrob.ghlint.model.effectivePermissions
+import net.twisterrob.ghlint.model.effectiveScopes
+import net.twisterrob.ghlint.rule.Example
+import net.twisterrob.ghlint.rule.Issue
+import net.twisterrob.ghlint.rule.Reporting
+import net.twisterrob.ghlint.rule.report
+import net.twisterrob.ghlint.rule.visitor.VisitorRule
+import net.twisterrob.ghlint.rule.visitor.WorkflowVisitor
+
+private typealias InferRequiredPermissions = (step: WorkflowStep.Uses) -> Set<RequiredScopes>
+
+private data class RequiredScopes(
+	val scope: Set<Scope>,
+	val reason: String,
+) {
+	constructor(reason: String, vararg scopes: Scope)
+			: this(@Suppress("detekt.SpreadOperator") setOf(*scopes), reason)
+
+	companion object {
+		val NO_GITHUB_TOKEN: Set<RequiredScopes> = setOf(
+			empty("No permissions are needed for the GitHub Token if a custom PAT is defined explicitly.")
+		)
+
+		fun empty(reason: String = "Not required."): RequiredScopes =
+			RequiredScopes(emptySet(), reason)
+	}
+}
+
+public class RequiredPermissionsRule : VisitorRule, WorkflowVisitor {
+	override val issues: List<Issue> = listOf(MissingRequiredActionPermissions)
+
+	override fun visitWorkflowUsesStep(reporting: Reporting, step: WorkflowStep.Uses) {
+		super.visitWorkflowUsesStep(reporting, step)
+
+		val definition = REQUIRED_PERMISSIONS_DEFINITIONS[step.uses.action] ?: return
+		val expectedPermissions = definition(step)
+		val effectivePermissions = step.parent.effectivePermissions ?: return
+		val definedPermissions = effectivePermissions.effectiveScopes
+
+		val remaining = expectedPermissions.flatMap { it.scope } - definedPermissions
```

Comment:
This is a little annoying as we then lose the ability to include the `reason` in the reporting.

Reply:
Yeah, I left it in half, hoping you would pick up, as you did 😁.